src/main/java/com/flansmod/common/driveables/mechas/EntityMecha.java

```diff
@@ -384,22 +389,13 @@ private void shoot(ItemStack stack, GunType gunType, ItemStack bulletStack, bool
 		bulletOrigin = Vector3f.add(new Vector3f(posX, posY, posZ), bulletOrigin, null);
 		
 		if(!world.isRemote)
-			for(int k = 0; k < gunType.numBullets * bulletType.numBullets; k++)
-			{
-				
-				// TODO: Do mechas properly. No hacks
-				float speed = gunType.getBulletSpeed(stack);
-				if(speed <= 0.0f)
-					speed = 5.0f;
-				world.spawnEntity(((ItemShootable)bulletStack.getItem()).getEntity(world,
-						bulletOrigin,
-						armVector,
-						(EntityLivingBase)(getSeat(0).getControllingPassenger()),
-						gunType.getSpread(stack) / 2F,
-						gunType.getDamage(stack),
-						speed,
-						mechaType));
-			}
+		{
+			ShootableType shootableType = ((ItemShootable)bulletStack.getItem()).type;
+			FireableGun fireableGun = new FireableGun(gunType, gunType.getDamage(stack), gunType.getSpread(stack), gunType.getBulletSpeed(stack));
+			//TODO unchecked cast, grenades will cause an error
```

Comment:
Mechas definitely can shoot grenades. I'm happy to put this PR in with some TODOs in though as we need to get some progress on it

Reply:
As far as my testing goes they can not use grenade launchers (#1023). I am currently unable to test this due to #1060, but i don't think anything changed